osd,crimson/osd: pass message using intrusive_ptr<>

tchaikov opened the pull request:
for two reasons:

* crimson::osd::PG::send_cluster_message() accepts a `Message*`
  pointer, and then hand it over to `shard_services.send_to_osd()`,
  which expects a `Ref<Message>`. so the raw pointer is used to
  construct an `intrusive_ptr<Message>`, which increment the
  refcount of that Message instance by one. but that Message
  was owned by nobody before that, so we end up with an
  `intrusive_ptr<Message>` of 2 refcount, and only a single
  owner. hence the memory leak.
* osd: to use Connection::send_message2(), which accepts
  MessageRef and less error-prone in the sense of preventing
  memory leak of Messages.

Signed-off-by: Kefu Chai <kchai@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

ronen-fr reviewed (approved):
Why not keep the original interface of the send_x_message_x functions in PG and in OSD, and
just add the new interface? then we can lazily change all calls.
(we can mark the old ones as [[deprecated]])

(I a mostly ref to your comment in #37303)

tchaikov commented:
these two interfaces are ambiguous if caller passes the raw pointer.

ronen-fr commented:
> these two interfaces are ambiguous if caller passes the raw pointer.

oopsy

tchaikov commented:
change since last revision:

```diff
diff --git a/src/crimson/osd/pg.h b/src/crimson/osd/pg.h
index 3b193697ae..2e9f6023cd 100644
--- a/src/crimson/osd/pg.h
+++ b/src/crimson/osd/pg.h
@@ -157,9 +157,9 @@ public:
   }

   void send_cluster_message(
     int osd, MessageRef m,
     epoch_t epoch, bool share_map_update=false) final {
-    (void)shard_services.send_to_osd(osd, MessageRef{m, false}, epoch);
+    (void)shard_services.send_to_osd(osd, m, epoch);
   }

   void send_pg_created(pg_t pgid) final {
```

tchaikov commented:
- https://pulpito.ceph.com/kchai-2020-09-26_11:23:24-rados-wip-kefu-testing-2020-09-25-2324-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-09-27_03:11:59-rados-wip-kefu-testing-2020-09-25-2324-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-09-27_14:40:39-rados-wip-kefu-testing-2020-09-25-2325-distro-basic-smithi

tchaikov commented:
- https://pulpito.ceph.com/kchai-2020-10-03_09:43:24-rados-wip-kefu-testing-2020-10-03-1611-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-10-04_04:38:32-rados-wip-kefu-testing-2020-10-03-1611-distro-basic-smithi/
